cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java

```diff
@@ -155,32 +167,85 @@ public void createNamespace(NamespaceMeta metadata) throws NamespaceCannotBeCrea
    * @throws NamespaceCannotBeDeletedException if the specified namespace cannot be deleted
    * @throws NotFoundException if the specified namespace does not exist
    */
-  public void deleteNamespace(Id.Namespace namespaceId) throws NamespaceCannotBeDeletedException, NotFoundException {
+  public void deleteNamespace(final Id.Namespace namespaceId)
+    throws NamespaceCannotBeDeletedException, NotFoundException {
     // TODO: CDAP-870, CDAP-1427: Delete should be in a single transaction.
-    if (store.getNamespace(namespaceId) == null) {
+    if (!hasNamespace(namespaceId)) {
       throw new NotFoundException(NAMESPACE_ELEMENT_TYPE, namespaceId.getId());
     }
-    // TODO: CDAP-870: Check if apps are running and abort if they are
-    // Delete Preferences associated with this namespace
-    preferencesStore.deleteProperties(namespaceId.getId());
 
-    // Delete all dashboards associated with this namespace
-    dashboardStore.delete(namespaceId.getId());
+    if (areProgramsRunning(namespaceId)) {
+      throw new NamespaceCannotBeDeletedException(namespaceId.getId(),
+                                                  "Some programs are currently running in namespace " + namespaceId);
+    }
 
-    // move data directory and try to remove HBase namespace
+    LOG.info("Deleting namespace '{}'.", namespaceId);
     try {
+      // Delete Preferences associated with this namespace
+      preferencesStore.deleteProperties(namespaceId.getId());
+      // Delete all dashboards associated with this namespace
+      dashboardStore.delete(namespaceId.getId());
+      // Delete datasets and modules
       dsFramework.deleteAllInstances(namespaceId);
       dsFramework.deleteAllModules(namespaceId);
-      dsFramework.deleteNamespace(namespaceId);
+      // Delete queues and streams data
+      queueAdmin.dropAllInNamespace(namespaceId.getId());
+      streamAdmin.dropAllInNamespace(namespaceId);
+      // Delete all meta data
+      store.removeAll(namespaceId);
+      // TODO: CDAP-1729 - Delete/Expire Metrics. API unavailable right now.
```

Comment:
The condition right below will ensure that the delete does not percolate to default namespaces in Hive, HBase, etc.